### PR TITLE
c# new var

### PR DESCRIPTION
**Pull request by gerhardol**
https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-9#fit-and-finish-features

Part of #8788, #8728
Follow up to #8729, covering more cases

Done after getting requests to change existing code in some PRs

## Proposed changes

Use C#9 new() handling and remove var types.

Submitted in commits with the regex I used to replace the types.
Some hundreds of manual changes too.
To be squashed.

This could be mostly "replayed" but I prefer to merge this now...

```
^(\s*(using )?)var(?<eq> \w+\s+=\s+)new\s+(((?<ga>GitArgumentBuilder)(?<gaa>.*[^;]*;$))|((?<in>\w+<.*>)(?<ina>\(\);$))|((?<st>\w+(<[^>]+>)?)(?<sta>(\([^\.]*\));?\s*$)))
 $1${ga}${in}${st}${eq}new${gaa}${ina}${sta}
Some incorrect changes manually converted

^(\s*(using )?)var(?<eq>\s+\w+\s+=\s+)new\s+(?<st>\w+(<[^>]+>)?)(?<sta>\s*$)
$1${st}${eq}new()${sta}

^(\s*(using )?)var(?<eq> \w+\s+=\s+)new\s+(((?<ga>GitArgumentBuilder)(?<gaa>.*[^;]*;$))|((?<in>\w+<.*>)(?<ina>\(\);$))|((?<st>\w+(<[^>]+>)?)(?<sta>(\(.*\));?\s*$)))
$1${ga}${in}${st}${eq}new${gaa}${ina}${sta}
Slightly to aggressive, some reverted

Manual changes, search using
^(\s*(using )?)var\s+(?<eq>\w+\s+=\s+)new\s+
```

## Test methodology 

Manual - currently little runtime but this should only be a reformatting

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Review by mstv — COMMENTED** [review]
Comprehensive - thank you

**Review by mstv — APPROVED** [review]

**Comment by gerhardol** [comment]
@msftbot merge in 4 days

**Comment by RussKie** [comment]
This will likely completely stall the migration to .NET 5. I'm already at the point of giving up due to all the conflicts I have to deal with...

**Comment by gerhardol** [comment]
> This will likely completely stall the migration to .NET 5. I'm already at the point of giving up due to all the conflicts I have to deal with...

This is not hard to repeat, the commit messages contains the changes done, except the final (not fixup) commit that was done manually (the search phrase is included, the change will require a couple of hours).
For this to work, the .NET5 branch must be rebased on the commit prior to this is merged. I can fix the rebase on this PR if you prefer.
If this is not merged, I expect several of these will contribute to changes later.

While this is quite large (about 1700 changes), the combination is smaller than the previous c#7/8/9 changes and that merge conflicts are much smaller than the other changes.

Feel free to stop the merge if you prefer.

Rebasing to resolve merge conflicts.

**Comment by RussKie** [comment]
Please base this work on .NET branch.

**Comment by gerhardol** [comment]
To be reapplied after .NET5

**Comment by gerhardol** [comment]
"Replayed" (regex and manual) the changes on master

File: GitUI/Editor/FileViewer.cs
Comment (drewnoakes) [inline GitUI/Editor/FileViewer.cs:1030]: Intentional?

File: GitUI/Editor/Diff/DiffLineNumAnalyzer.cs
Comment (drewnoakes) [inline GitUI/Editor/Diff/DiffLineNumAnalyzer.cs:19]: Discussion point: For projects that prefer var, I think a change like this makes the code look a bit inconsistent, as the variables no longer line up. It's a minor point, and of course any code style can be learned, but it does seem a bit harder to read with this change.

I feel much more strongly about the case where the type name is completely removed (e.g. `foo.Bar(new())`), but that does not appear to be present in this PR.

File: GitUI/Editor/FileViewer.cs
Comment (drewnoakes) [reply GitUI/Editor/FileViewer.cs:1030]: Also, was there not a Roslyn codefix to make these changes? If one exists, why didn't you use it? If it has issues, I'd like to report them.

Comment (gerhardol) [reply GitUI/Editor/FileViewer.cs:1030]: This was intentional. new() cannot be used here and I reverted the change. In this case I followed IDE0008 and kept the type. Will remove that from this PR.

It is not possible to apply IDE0090 without removing `var` first. So IDE0008 has to be applied which removes all var declarations before IDE0090 can be applied.
Adding a commit to apply IDE0090, there are some usages, especially in definitions.
Who could imagine a compiler was better than a regex?

File: GitUI/Editor/Diff/DiffLineNumAnalyzer.cs
Comment (drewnoakes) [reply GitUI/Editor/Diff/DiffLineNumAnalyzer.cs:19]: I like the changes in e1168fb4885c7c1d0b649c972551a71c7aa2593a.

Comment (RussKie) [reply GitUI/Editor/Diff/DiffLineNumAnalyzer.cs:19]: I don't like `var` because more often than not it makes the code opaque (e.g. `var result = GetSomething();`), and prefer to avoid it.
The new `new` syntax makes it better in my opinion.

Also I think I rather like `foo.Bar(new())` instead of `foo.Bar(new Some.CrazyLongTypeName.ThatIDontCareAbout())` because it provides enough information for me. Obviously, like with anything else, a common sense must be applied as this may become unreadable `foo.Bar(new(), new(new, new()), new())`

**Comment by RussKie** [comment]
> Should I change this to apply IDE0008 (remove all var declarations) followed by IDE0090 (new())?

Sure, if you think it's worth it.

I'm happy for this to merge (once the build is green), and then have a follow up.

**Comment by gerhardol** [comment]
> > Should I change this to apply IDE0008 (remove all var declarations) followed by IDE0090 (new())?
> 
> Sure, if you think it's worth it.
> 
> I'm happy for this to merge (once the build is green), and then have a follow up.

I personally am no fan of removing all var, type is often obvious and I prefer shorter code.
But if IDE0008 is the preference it may be better to apply it universally.

A little surprised that Codefactor found 8 fixed issues and 10 new after this change...

Would like to squash and merge this PR a couple of days after next approval.

File: GitUI/HelperDialogs/FormResetCurrentBranch.cs
Comment (mstv) [inline GitUI/HelperDialogs/FormResetCurrentBranch.cs:27]: This omission is located at the edge of readability *for me*. Not requesting a change.

File: GitUI/Editor/Diff/DiffLineNumAnalyzer.cs
Comment (drewnoakes) [reply GitUI/Editor/Diff/DiffLineNumAnalyzer.cs:19]: It seems contradictory to dislike var due to invisible types, while also preferring new Foo(new())). Am I missing something?

One issue with new Foo(new())) is that if you add an overload of Foo, your code no longer compiles.

Comment (gerhardol) [reply GitUI/Editor/Diff/DiffLineNumAnalyzer.cs:19]: > It seems contradictory to dislike var due to invisible types, while also preferring new Foo(new())). Am I missing something?

I find something like the following fine, then the type is "close".
```c#
List<DateTime> l = new() { new() };
```

> One issue with new Foo(new())) is that if you add an overload of Foo, your code no longer compiles.

Need to be handled by the designer adding the overload or updating a library.

But I do not want to use `new()` much more than in this PR or really remove `var` generally.

**Review by mstv — APPROVED** [review]
Thank you

---

> A little surprised that Codefactor found 8 fixed issues and 10 new after this change...

Codefactor seems to be commenting every touched function regardless of the change (seen in other PR, too).

File: GitCommands/Git/ExecutableExtensions.cs
Comment (mstv) [reply GitCommands/Git/ExecutableExtensions.cs:120]: For another PR: applies at least to `MemoryStream` and `*Form` IIUIC.

File: GitUI/HelperDialogs/FormResetCurrentBranch.cs
Comment (gerhardol) [reply GitUI/HelperDialogs/FormResetCurrentBranch.cs:27]: I see your point in this case, even I prefer the change also here.
It is hard to have consistent rules without reasonable exceptions. I do not really want IDE0008 to be enforced for instance but to keep it as a recommendation where needed.

**Comment by gerhardol** [comment]
https://github.com/gitextensions/gitextensions/pull/9182#discussion_r631432160

Adding warning if not followed (no compile error)
image

File: GitUI/Editor/Diff/DiffLineNumAnalyzer.cs
Comment (RussKie) [reply GitUI/Editor/Diff/DiffLineNumAnalyzer.cs:19]: > It seems contradictory to dislike var due to invisible types, while also preferring new Foo(new())). Am I missing something?

I see no contradictions here.
I want to see what type is being returned from a call so I can reason about how it is used. But when I'm see a type constructed I don't _really_ care what goes in (the compiler will check).

File: GitCommands/Git/ExecutableExtensions.cs
Comment (drewnoakes) [reply GitCommands/Git/ExecutableExtensions.cs:120]: Disposing a `MemoryStream` does nothing, so this change is technically a pessimisation. But it's not a big deal either way.

**Comment by gerhardol** [comment]
Squashed with merge conflict resolve on master. Left a tmp/ branch behind
Set to merge (will need a conflict resolve rebase I assume) in four days.

**Pull request merged**
